### PR TITLE
[Markdown] added doc string for @md_str string literal

### PR DESCRIPTION
**Pull request by mitiemann**
This PR addresses #51168 . I guess this PR wants to have the labels doc & markdown .

### CODE REVIEW

**Comment by mitiemann** [comment]
Hello, world! This is my first PR to julia, some I'd be grateful for feedback. In particular:

- [ ] Maybe it would be nice adding an example with a `jldoctest`. However, as `md"# Hello, world!"` is pretty-printed in the REPL, I'm wondering how to do this elegantly? Maybe wrapping it in an HTML renderer?
- [ ] I've noticed `@doc_str` also doesn't have a docstring as of yet. However, I'm wondering whether this should really be part of the exposed API?

**Comment by mitiemann** [comment]
There is a simultaneous PR for this issue at #52607 . We're currently figuring out which aspect of which PR to keep.

File: stdlib/Markdown/src/Markdown.jl [hunk 1]
```diff
@@ -40,6 +40,11 @@ function docexpr(source::LineNumberNode, mod::Module, s, flavor = :julia)
     :($doc_str($(mdexpr(s, flavor)), $(QuoteNode(source)), $mod))
 end
 
+"""
+    @md_str -> MD
+
+Construct a `MD` Markdown object from the given string.
```
Comment (stevengj) [inline stdlib/Markdown/src/Markdown.jl:46]:
```suggestion
Parse the given string as Markdown text and return a corresponding [`Markdown.MD`](@ref) object.
```

Comment (mitiemann) [reply stdlib/Markdown/src/Markdown.jl]: Thanks! I've used your wording, but I've removed the `(@ref)` as this didn't compile. I'm assuming that this would have required to also add the `Markdown.MD` type into the docs which seemed excessive to me.

File: NEWS.md [hunk 2]
```diff
@@ -111,6 +115,13 @@ Standard library changes
...
    respectively, now efficiently compute the generalized eigenvalues (`eigen`: and eigenvectors) of `A` and `B`. Note: The second
    argument is the output of `bunchkaufman` or `lu` ([#50471]).
+* Structured matrices now retain either the axes of the parent (for `Symmetric`/`Hermitian`/`AbstractTriangular`/`UpperHessenberg`), or that of the principal diagonal (for banded matrices) ([#52480]).
+* `bunchkaufman` and `bunchkaufman!` now work for any `AbstractFloat`, `Rational` and their complex variants. `bunchkaufman` now supports `Integer` types, by making an internal conversion to `Rational{BigInt}`. Added new function `inertia` that computes the inertia of the diagonal factor given by the `BunchKaufman` factorization object of a real symmetric or Hermitian matrix. For complex symmetric matrices, `inertia` only computes the number of zero eigenvalues of the diagonal factor ([#51487]).
+
+#### Logging
+* New `@create_log_macro` macro for creating new log macros like `@info`, `@warn` etc. For instance
+  `@create_log_macro MyLog 1500 :magenta` will create `@mylog` to be used like `@mylog "hello"` which
+  will show as `┌ MyLog: hello` etc. ([#52196])
```
Comment (stevengj) [inline NEWS.md:124]: It seems like you messed up a merge somehow and pulled a whole bunch of unrelated changes into this PR?

Comment (mitiemann) [reply NEWS.md:124]: Yeah, I've only wanted to rebase to the current changes in base, but I was in a hurry and then ...
Is there anything else I need to fix before this can be merged?

**Comment by mitiemann** [comment]
Okay, I believe this would be ready to merge @stevengj . This is my first Julia PR. Is there anything else from my side that I could help to merge this?

File: stdlib/Markdown/src/Markdown.jl [hunk 3]
```diff
@@ -40,6 +40,18 @@ function docexpr(source::LineNumberNode, mod::Module, s, flavor = :julia)
...
 end
 
+"""
+    macro md_str(s)
+
+Parse the given string as Markdown text and return a corresponding `Markdown.MD` object.
+
+# Example
+```jldoctest
+julia> html(md"# Hello, world!")
```
Comment (stevengj) [inline stdlib/Markdown/src/Markdown.jl:50]: Note that the `html` function also doesn't have a docstring…

Can we just show the example as
```jl
julia> md"# Hello, world!"
  Hello, world!
  ≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡
```
similar to the way it appears in the REPL with `julia --color=no`?

Comment (mitiemann) [reply stdlib/Markdown/src/Markdown.jl:50]: > Note that the `html` function also doesn't have a docstring 😢 …

I can give that a try as well (and generally have a look at the exported API). I'm wondering whether this would also warrant a more general documentation about how this can be used to export Markdown to various formats? I will have a general look.

> Can we just show the example as
> ```jl
> julia> s = md"# Hello, world!"
>   Hello, world!
>   ≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡
> 
> julia> typeof(s)
> Markdown.MD
> ```
> similar to the way it appears in the REPL with `julia --color=no`? 

I tried this at some point but the bold formatting(?) gave me problems with the `doctest` failing after copy-and-pasting. I can try to see whether I can get it working after all. 

>  We could also add `show` calls, which *are* documented:
> ```jl
> julia> show(stdout, "text/markdown", s)
> # Hello, world!
> 
> julia> show(stdout, "text/html", s)
> <div class="markdown"><h1>Hello, world&#33;</h1>
> </div>
> 
> julia> show(stdout, "text/plain", s)
>   Hello, world!
>   ≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡
> ```
> though it might be better to put those into a docstring for `Markdown.MD`.

This could be another option. I will play around a bit.

Thanks for your swift reply. I'll ping you again once I'm satisfied.

Comment (fonsp) [reply stdlib/Markdown/src/Markdown.jl:54]: Docstring for `html` added in #52733 ☺️

File: stdlib/Markdown/src/Markdown.jl [hunk 4]
```diff
@@ -1,7 +1,11 @@
 # This file is a part of Julia. License is MIT: https://julialang.org/license
 
 """
+    Markdown
+
 Tools for working with the Markdown file format. Mainly for documentation.
+The `Markdown` module provides the (internal) `MD` struct (type?) as well as the string
```
Comment (fonsp) [inline stdlib/Markdown/src/Markdown.jl:7]:
```suggestion
The `Markdown` module provides the (internal) `MD` type as well as the string 
```

File: stdlib/Markdown/src/Markdown.jl [hunk 5]
```diff
@@ -40,6 +44,18 @@ function docexpr(source::LineNumberNode, mod::Module, s, flavor = :julia)
     :($doc_str($(mdexpr(s, flavor)), $(QuoteNode(source)), $mod))
 end
 
+"""
+    macro md_str(s)
```
Comment (fonsp) [inline stdlib/Markdown/src/Markdown.jl:48]:
```suggestion
    @md_str -> Markdown.MD
```

to match the convention used here https://docs.julialang.org/en/v1/base/strings/#Base.Docs.@html_str

File: stdlib/Markdown/src/Markdown.jl (hunk 3 above)
Comment (fonsp) [reply stdlib/Markdown/src/Markdown.jl:54]: I really like the example with `s = md"# Heading`, `typeof(s)`. Maybe we can just use this example without doctest to avoid the bold formatting problem?

File: stdlib/Markdown/src/Markdown.jl [hunk 6]
```diff
@@ -51,6 +67,12 @@ function doc_str(md, source::LineNumberNode, mod::Module)
 end
 doc_str(md::AbstractString, source::LineNumberNode, mod::Module) = doc_str(parse(md), source, mod)
 
+"""
+    macro doc_str(s)
+
+Parse the given string as Markdown text, add line and module information and return a 
+corresponding `Markdown.MD` object.
```
[the remark this answers is not in the document]
Comment (fonsp) [reply stdlib/Markdown/src/Markdown.jl:74]: To answer your question:

> I've noticed @doc_str also doesn't have a docstring as of yet. However, I'm wondering whether this should really be part of the exposed API?

We cannot make a breaking change, so the `@doc_str` needs to stay exported. ([It's already used in packages.](https://juliahub.com/ui/Search?q=doc%22&type=code)) We try to have the convention that all exported fields are documented.

[the remark this answers is not in the document]
Comment (fonsp) [reply stdlib/Markdown/src/Markdown.jl:74]: You could also write about the usage in `Base.@doc doc"New docstring" existing_definition`

File: stdlib/Markdown/src/Markdown.jl (hunk 5 above)
Comment (mitiemann) [reply stdlib/Markdown/src/Markdown.jl:48]: I've copied the convention I found in [`Artifacts.@artifact_str`](https://docs.julialang.org/en/v1/stdlib/Artifacts/#Artifacts.@artifact_str). If I find the time, I'll try to find all (string literal) macro docs, check for their conventions and file an issue if I find too many divergent styles.

File: stdlib/Markdown/src/Markdown.jl (hunk 3 above)
Comment (mitiemann) [reply stdlib/Markdown/src/Markdown.jl]: This now magically works on my machine, so I've reverted to `s = md"# Hello, world!"` and hope it won't break the build

**Comment by mitiemann** [comment]
Thanks :-) Should I squash the commits myself or is this handled during merging? Also, only merge after #52733 has been merged as `index.md` now also includes `Markdown.html` and `Markdown.latex`.

**Comment by stevengj** [comment]
Github will squash for you, so you needn't bother.

File: stdlib/Markdown/src/parse/parse.jl [hunk 7]
```diff
@@ -1,5 +1,11 @@
 # This file is a part of Julia. License is MIT: https://julialang.org/license
 
+"""
+    MD
+
+`MD` represents a Markdown document. However, `MD` objects should only be constructed using 
+the exported macros [`@md_str`](@ref) and [`@doc_str`](@ref).
+"""
 mutable struct MD
```
Comment (stevengj) [inline stdlib/Markdown/src/parse/parse.jl:9]: We should also have a `public MD` statement if this is a part of the documented API.

File: stdlib/Markdown/src/parse/parse.jl [hunk 8]
```diff
@@ -1,5 +1,11 @@
 # This file is a part of Julia. License is MIT: https://julialang.org/license
 
+"""
+    MD
+
+`MD` represents a Markdown document. However, `MD` objects should only be constructed using
```
Comment (stevengj) [inline stdlib/Markdown/src/parse/parse.jl]:
```suggestion
`MD` represents a Markdown document.  Note that the `MD(...)` constructor should not generally be used
directly, since it constructs the internal data structures.   Instead, you can construct `MD` objects using
```
Good here to be explicit that we are warning against calling the `MD(...)` constructor directly.

For example, there are also the `Markdown.parse` functions that we might want to document at some point.

Alternatively, we might want to define:
```jl
MD(md::AbstractString) = parse(md)
```
since as far as I can tell there is no need to ever call the constructor with a single string argument — you generally want to to have at least a `Markdown.Paragraph` object, and in any case you can always call the lower-level constructor `MD([str])` in the event that you really do want the `content` to consist of a single string.  However, this is a change that should go in a separate PR.

File: stdlib/Markdown/src/Markdown.jl [hunk 9]
```diff
@@ -1,7 +1,11 @@
 # This file is a part of Julia. License is MIT: https://julialang.org/license
 
 """
+    Markdown
+
 Tools for working with the Markdown file format. Mainly for documentation.
```
Comment (stevengj) [inline stdlib/Markdown/src/Markdown.jl:6]:
```suggestion
Tools for working with the Markdown markup language for formatted text, used within Julia for documentation.
```

File: stdlib/Markdown/src/parse/parse.jl (hunk 8 above)
Comment (mitiemann) [reply stdlib/Markdown/src/parse/parse.jl]: I agree with documenting the `Markdown.parse` function, although I'm guessing that this should then include a description of the entire parse logic, including the `Markdown.@flavor` macro. I'm just not sure whether this should be part of the public API or part of the developer documentation.

Comment (stevengj) [reply stdlib/Markdown/src/parse/parse.jl]: If we are providing a Markdown module, to me it's pretty basic functionality to document how to parse strings generated at runtime (as opposed to literal strings).  This isn't a low-level thing, like the internals of the `MD` type.

That being said, I agree that the flavors open a can of worms, so it might be better to leave this for a separate issue/PR so that this one can be merged.

File: stdlib/Markdown/src/parse/parse.jl [hunk 10]
```diff
@@ -1,5 +1,12 @@
 # This file is a part of Julia. License is MIT: https://julialang.org/license
 
+"""
+    MD
+
+`MD` represents a Markdown document. Note that the `MD` constructor should not generally be
+used directly, since it constructs the internal data structures.   Instead, you can
```
Comment (waldyrious) [inline stdlib/Markdown/src/parse/parse.jl:7]: Is the use of triple spaces to separate the sentences intentional? It doesn't seem to be used elsewhere in this PR. :thinking:

Comment (stevengj) [reply stdlib/Markdown/src/parse/parse.jl:7]: Probably an accident, though some people reflexively type two spaces after every sentence.  In any case, it's harmless, since Julia's Markdown parser ignores the additional whitespace:
```jl
julia> md"Foo.      Bar."
  Foo. Bar.
```

**Comment by stevengj** [comment]
CI failures seem unrelated.

**Comment by stevengj** [comment]
CI failures are unrelated (#52506 etc.)
